### DISCUSSION TITLE
Binary installer request

### DISCUSSION THREAD

**Discussion by JimSEvans** [thoughtspot/cs_tools#73] [General] [answered]
Hi there,

Could I please get a binary installer?
```
# python -m sysconfig
Platform: "linux-x86_64"
Python version: "3.11"
Current installation scheme: "posix_prefix"
```
Thanks!
Jim

**Comment by boonhapus** [comment, marked as answer]
Hey @JimSEvans ! Apologies for the slow response on this one.

1. Extract the zip file.
2. Run python against the `_bootstrapper.py` file with `--offline-mode` specified

`cd cs-tools_1.4.12_linux-x86_64_3.11`
`python _bootstrapper.py --install --offline-mode .`

It's a bit too big for GitHub discussions, so I've attached an Egnyte link. Let me know if that works for you!

[https://thoughtspot.egnyte.com/dl/RW08FEYYAV/cs-tools_1.4.12_linux-x86_64_3.11.zip_](https://thoughtspot.egnyte.com/dl/RW08FEYYAV/cs-tools_1.4.12_linux-x86_64_3.11.zip_)

**Reply by JimSEvans** [reply]
Thanks! Sorry but I think I missed the window-- the link isn't working for me.

**Reply by boonhapus** [reply]
@JimSEvans the hyperlink is just weird. It's got a trailing `_` that GitHub was chopping off. Can you try it again?

**Reply by JimSEvans** [reply]
Hi @boonhapus, thanks again for the installer. But I noticed it seems to be lacking Cython<3.0. I've tried installing within a running container, and via Dockerfile and get the same error installing Cython, even if I pip install cython (version less than 3.0) prior to running ``python _bootstrapper.py --install --offline-mode .``, I still get the error. I don't suppose there's any way you could include all dependencies in the installer?

**Reply by boonhapus** [reply]
@JimSEvans The problem appears to be **[pyyaml#736](https://github.com/yaml/pyyaml/issues/736)** / **[v6.0.1 compatability](https://github.com/yaml/pyyaml/commit/c42fa3bff1eabdb64763bb1526d9ea1ccb708479)** on Linux builds. We use `pyyaml >= 6.0` in the upstream **[`thoughtspot_tml` library](https://github.com/thoughtspot/thoughtspot_tml/blob/a9d6715c3e9a1b345121dbbbae91a202a544a763/pyproject.toml#L34C21-L34C21)** and I'm not interested in supporting `5.4.1` (the previous major version).

I'll have to explore ways to get this to work on the cluster / Linux systems. I **_think_** I should be able to build the wheel programmatically and supply it, but need to test this. I'll try this today.

**Reply by JimSEvans** [reply]
Understood. I hugely appreciate it!

**Reply by boonhapus** [reply]
@JimSEvans are you able to try the same Egnyte link again? I have manually added a few of the binaries, so this shouldn't need to be rebuilt with `Cython` anymore. This _should_ fix your issue.

It'll pose problems for my build process later.. but that's a problem for future me. 😅👍

**Reply by JimSEvans** [reply]
Thanks! With your updated folder, I just needed to add lines to my dockerfile to yum install gcc and python3.11-devel and it installed. Running "cs_tools --version" worked so I think I'm good to go. I greatly appreciate the time you put into helping me out!

**Reply by boonhapus** [reply]
In fact @JimSEvans ... that might have solved your issue in the first place LOL

The only thing I did was grab the .whl from pypi instead of the uncompiled binary. My offline deployment process essentially vendors `pip download --python-version X.Y --platform PLAT_NAME --no-binary` because _most platforms_ include Python's development headers by default. The CS Tools user base is primarily internet-connected installs (98% of 125 customers), but also primarily Windows (~70% of 125 customers).

I'm curious -- what does your process look like? I'm surprised you're deploying via docker but also need an offline binary? 🤔
